core/lib/zksync_core/src/api_server/web3/state.rs

```diff
@@ -182,6 +184,58 @@ impl RpcState {
         }
     }
 
+    /// Returns an error if the provided `block` is known to be pruned.
+    pub(crate) fn check_pruned_block(&self, block: api::BlockId) -> Result<(), Web3Error> {
+        let first_local_miniblock = self.start_info.first_miniblock;
+        match block {
+            api::BlockId::Number(api::BlockNumber::Number(number))
+                if number < first_local_miniblock.0.into() =>
+            {
+                Err(Web3Error::PrunedBlock(first_local_miniblock))
+            }
+            api::BlockId::Number(api::BlockNumber::Earliest)
+                if first_local_miniblock > MiniblockNumber(0) =>
+            {
+                Err(Web3Error::PrunedBlock(first_local_miniblock))
+            }
+            _ => Ok(()),
+        }
+    }
+
+    pub(crate) fn check_pruned_l1_batch(&self, number: L1BatchNumber) -> Result<(), Web3Error> {
+        let first_local_l1_batch = self.start_info.first_l1_batch;
+        if number < first_local_l1_batch {
+            return Err(Web3Error::PrunedL1Batch(first_local_l1_batch));
+        }
+        Ok(())
+    }
```

Comment:
Idk if it'll be better, but as a suggestion:
If we will move `BlockStartInfo` to the `api_server`, we can in theory do roughly the following (pseudocode):

```rust
enum PruneQuery {
    Batch(L1BatchNumber),
    ApiBlock(api::BlockId),
}

impl BlockStartInfo {
  fn ensure_not_pruned(&self, query: impl Into<PruneQuery>) -> Result<(), Web3Error> { .. }
}
```

then you'll be able to just call `self.state.block_start_info.ensure_not_pruned(whatever)?`.